rust/lance-index/src/scalar/expression.rs

```diff
@@ -388,31 +452,106 @@ impl ScalarIndexExpr {
     /// any situations where the session cache has been disabled.
     #[async_recursion]
     #[instrument(level = "debug", skip_all)]
-    pub async fn evaluate(&self, index_loader: &dyn ScalarIndexLoader) -> Result<RowIdMask> {
+    pub async fn evaluate(&self, index_loader: &dyn ScalarIndexLoader) -> Result<IndexExprResult> {
         match self {
             Self::Not(inner) => {
                 let result = inner.evaluate(index_loader).await?;
-                Ok(!result)
+                match result {
+                    IndexExprResult::Exact(mask) => Ok(IndexExprResult::Exact(!mask)),
+                    IndexExprResult::AtMost(mask) => Ok(IndexExprResult::AtLeast(!mask)),
+                    IndexExprResult::AtLeast(mask) => Ok(IndexExprResult::AtMost(!mask)),
+                }
             }
             Self::And(lhs, rhs) => {
                 let lhs_result = lhs.evaluate(index_loader);
                 let rhs_result = rhs.evaluate(index_loader);
                 let (lhs_result, rhs_result) = join!(lhs_result, rhs_result);
-                Ok(lhs_result? & rhs_result?)
+                match (lhs_result?, rhs_result?) {
+                    (IndexExprResult::Exact(lhs), IndexExprResult::Exact(rhs)) => {
+                        Ok(IndexExprResult::Exact(lhs & rhs))
+                    }
+                    (IndexExprResult::Exact(lhs), IndexExprResult::AtMost(rhs))
+                    | (IndexExprResult::AtMost(lhs), IndexExprResult::Exact(rhs)) => {
+                        Ok(IndexExprResult::AtMost(lhs & rhs))
+                    }
+                    (IndexExprResult::Exact(lhs), IndexExprResult::AtLeast(_)) => {
+                        // We could do better here, elements in both lhs and rhs are known
+                        // to be true and don't require a recheck.  We only need to recheck
+                        // elements in lhs that are not in rhs
```

Comment:
it seems like we almost want these IndexExprResults to be sets of enums. That would also let us handle cases like negation in the new ngram index, where

```
select * from foo where not contains(t, "abcde")
```

currently would get a full recheck, but if we recognized the negation and served it from the index, we'd only need to recheck the rows that had overlapping trigrams and could answer exactly for the others.

Reply:
Yes.  Right now we return:

```
{
  block_list: RowIdTreeMap,
  allow_list: RowIdTreeMap,
}
```

I think we eventually want something like:

```
{
  definitely_false: RowIdTreeMap,
  likely_false: RowIdTreeMap,
  likely_true: RowIdTreeMap,
  definitely_true: RowIdTreeMap,
}
```

However, we don't really have the machinery to process that kind of result.  I was thinking this could be tackled around the same time after we tackle the "index size too large for a take" scenario.  Let's sync on this today.